google/cloud/bigquery/query.py

```diff
@@ -186,31 +186,56 @@ class ArrayQueryParameter(_AbstractQueryParameter):
 
         array_type (str):
             Name of type of array elements.  One of `'STRING'`, `'INT64'`,
-            `'FLOAT64'`, `'NUMERIC'`, `'BOOL'`, `'TIMESTAMP'`, or `'DATE'`.
-
-        values (List[appropriate scalar type]): The parameter array values.
+            `'FLOAT64'`, `'NUMERIC'`, `'BOOL'`, `'TIMESTAMP'`, `'DATE'`,
+            or `'STRUCT'`/`'RECORD'`.
+
+        values (List[appropriate type]): The parameter array values.
+
+        struct_item_type (Optional[google.cloud.bigquery.query.StructQueryParameter]):
```

Comment:
Could this be an iterable of [SchemaField](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.schema.SchemaField.html#google.cloud.bigquery.schema.SchemaField)? Since we only want type information, not actual values, it seems the query parameter classes are the wrong choice for this argument.

Reply:
Are those serialized into a compatible API representation and the backend won't notice? If yes, then let's do that.

I suppose tests should sufficiently protect us if these two representations drift apart at some point.

Reply:
Good point. They aren't 100% the same. No "mode" in the query parameter types because it's following the standard SQL semantics. Instead of this implementation, perhaps we should introduce a [QueryParameterType](https://cloud.google.com/bigquery/docs/reference/rest/v2/QueryParameter#QueryParameterType) to match the API backend?

Then, `array_type` could take `Union[str, QueryParameterType]` instead of just `str`.

Reply:
Both using a `*QueryParameter` instance with an empty value is or a `SchemaField` instance are a bit of a hack that might eventually bite use, thus introducing `QueryParameterType` sounds like the "most correct" solution, but it will require some extra work.

If that's fine, I can take a stab, WDYT?